rest/diagnostic_api.go

```diff
@@ -0,0 +1,158 @@
+					} else {
+						resp.DynamicRoleGrants[roleName][keyspace][channel] = chanHistory
+					}
+				}
+			}
+		}
+	}
+
+	// Loop over previous roles, no way to determine admin and dynamic atm
+	for roleName, roleHist := range user.RoleHistory() {
+		role, err := h.db.Authenticator(h.ctx()).GetRole(roleName)
+		if err != nil {
+			return err
+		}
+		if role == nil {
+			continue
+		}
+		collAccessAll := role.GetCollectionsAccess()
+		resp.RoleHistoryGrants[roleName] = make(map[string]map[string]auth.GrantHistory)
+
+		for scopeName, collections := range collAccessAll {
+			for collectionName, collectionAccess := range collections {
+				keyspace := scopeName + "." + collectionName
+				resp.RoleHistoryGrants[roleName][keyspace] = make(map[string]auth.GrantHistory)
+				// loop over current role channels
+				for channel, _ := range collectionAccess.Channels() {
+					resp.RoleHistoryGrants[roleName][keyspace][channel] = roleHist
+				}
+				// loop over previous role channels
+				for channel, chanHistory := range collectionAccess.ChannelHistory() {
+					if chanHistory.Entries[len(chanHistory.Entries)-1].StartSeq < roleHist.Entries[len(roleHist.Entries)-1].StartSeq {
+						chanHistory.Entries[len(chanHistory.Entries)-1].StartSeq = roleHist.Entries[len(roleHist.Entries)-1].StartSeq
+					}
+					resp.RoleHistoryGrants[roleName][keyspace][channel] = chanHistory
+				}
+			}
+		}
+	}
+
+	// Loop over current and past channels
+	for scope, collectionConfig := range user.GetCollectionsAccess() {
+		for collectionName, CAConfig := range collectionConfig {
+			keyspace := scope + "." + collectionName
+			resp.AdminGrants[keyspace] = make(map[string]auth.GrantHistory)
+			resp.DynamicGrants[keyspace] = make(map[string]auth.GrantHistory)
+			// current channels
+			for channel, seq := range CAConfig.Channels() {
+				var history auth.GrantHistory
+				// If channel is in history, copy grant history
+				if _, ok := CAConfig.ChannelHistory_[channel]; ok {
+					history = CAConfig.ChannelHistory_[channel]
+					// Else, assign sequence channel was granted as startSeq on new grant history
+				} else {
+					history = auth.GrantHistory{Entries: []auth.GrantHistorySequencePair{{StartSeq: seq.Sequence}}}
+				}
+
+				if CAConfig.ExplicitChannels_.Contains(channel) {
+					resp.AdminGrants[keyspace][channel] = history
+				} else {
+					resp.DynamicGrants[keyspace][channel] = history
+				}
+			}
+			for channel, chanHistory := range CAConfig.ChannelHistory() {
+				if chanHistory.AdminAssigned {
+					resp.AdminGrants[keyspace][channel] = CAConfig.ChannelHistory_[channel]
+				} else {
+					resp.DynamicGrants[keyspace][channel] = CAConfig.ChannelHistory_[channel]
+				}
+			}
+		}
+	}
+
+	if !h.db.OnlyDefaultCollection() {
+		bytes, err := base.JSONMarshal(resp)
+		h.writeRawJSON(bytes)
+		return err
+	}
+	info := marshalPrincipal(h.db, user, true)
+	bytes, err := base.JSONMarshal(info.Channels)
+	h.writeRawJSON(bytes)
+	return err
```

Comment:
This looks like an unhandled error if we're writing a response before returning `err`.